### PR TITLE
Expose `rb_str_append` as `String#append` and special case BINARY encoded buffers.

### PR DESCRIPTION
**Pull request by ioquatix**
Please see https://bugs.ruby-lang.org/issues/14975 for discussion.

### CODE REVIEW

**Comment by ioquatix** [comment]
@nobu @ko1 do you mind giving me feedback? Thanks :)

File: string.c [hunk 1]
```diff
@@ -2927,6 +2927,22 @@ rb_str_append(VALUE str, VALUE str2)
     return rb_str_buf_append(str, str2);
 }
 
+
+VALUE
+rb_str_append2(VALUE str, VALUE str2)
+{
+    StringValue(str2);
+    
+    if (RB_ENCODING_IS_ASCII8BIT(str) || ENCODING_GET(str) == ENCODING_GET(str2)) {
```
Comment (eregon) [inline string.c]: That check is too simple, it should be something like "is str encoding's a superset of str2's encoding?".
e.g. it's fine to append US-ASCII to UTF-8.

Comment (eregon) [reply string.c]: Or alternatively make this method only work on binary strings and name it `byteappend` (see issue).

**Comment by ioquatix** [comment]
`IO::Buffer` makes this less important.

**Pull request closed without merging**